### PR TITLE
Centralize command running in boostrap (part one)

### PR DESCRIPTION
**Pull request by Kobzol**
This PR tries to consolidate the various `run, try_run, run_quiet, run_quiet_delaying_failure, run_delaying_failure` etc. methods on `Builder`. This PR only touches command execution which doesn't produce output that would be later read by bootstrap, and it also only refactors spawning of commands that happens after a builder is created (commands executed during download & git submodule checkout are left as-is, for now).

The `run_cmd` method is quite meaty, but I expect that it will be changing rapidly soon, so I considered it easy to kept everything in a single method, and only after things settle down a bit, then maybe again split it up a bit.

I still kept the original shortcut methods like `run_quiet_delaying_failure`, but they now only delegate to `run_cmd`. I tried to keep the original behavior (or as close to it as possible) for all the various commands, but it is a giant mess, so there may be some deviations. Notably, `cmd.output()` is now always called, instead of just `status()`, which was called previously in some situations.

Apart from the refactored methods, there is also `Config::try_run`, `check_run`, methods that run commands that produce output, oh my… that's left for follow-up PRs :)

The driving goal of this (and following) refactors is to centralize command execution in bootstrap on a single place, to make command mocking feasible. 

r? @onur-ozkan

### CODE REVIEW

**Comment by onur-ozkan** [comment]
Fails on CI currently @rustbot author

File: src/bootstrap/lib.rs
Comment (onur-ozkan) [inline src/bootstrap/lib.rs:995]: Shouldn't we handle this in `exec` module?

Comment (Kobzol) [reply src/bootstrap/lib.rs:995]: That's not so easy yet - we need to access `self.fail_fast`, `self.delayed_failures`, `self.verbose` etc. It would need more refactoring to introduce some shared command context or something like that.

File: src/bootstrap/lib.rs
Comment (onur-ozkan) [inline src/bootstrap/lib.rs:46]: Can you group `pub use` and `use` separately?

File: src/bootstrap/lib.rs
Comment (onur-ozkan) [inline src/bootstrap/lib.rs:1055]: This seems very confusing. When exactly `command.failure_behavior` is `BehaviorOnFailure::DelayFail` and `self.fail_fast` is true ? :thinking:

File: src/bootstrap/exec.rs
Comment (onur-ozkan) [inline src/bootstrap/exec.rs:28]: All `BehaviorOnFailure` does is either exit on failure, or keep the execution until the end. In this case `fail_fast: bool` should be able to replace `Option<BehaviorOnFailure>` here ?

File: src/bootstrap/exec.rs
Comment (onur-ozkan) [inline src/bootstrap/exec.rs:21]:
```suggestion
    SuppressOnSuccess,
```

File: src/bootstrap/test.rs
Comment (onur-ozkan) [inline src/bootstrap/test.rs:3092]: Previous execution ensures the success of the command with `unwrap`.

**Review by onur-ozkan — CHANGES_REQUESTED** [review]

File: src/bootstrap/lib.rs
Comment (Kobzol) [reply src/bootstrap/lib.rs:1055]: `failure_behavior` is the property of a command, and it is decided by the programmer who implemented that command's execution.

`fail_fast` is a property of a bootstrap CLI subcommand which states whether this behavior should be overridden to exit immediately. For the test subcommand, fail fast is the default, but users of bootstrap can specify on the CLI that they want to turn this behavior off.

File: src/bootstrap/exec.rs
Comment (Kobzol) [reply src/bootstrap/exec.rs:28]: Good point. I didn't realize that when this is `None`, the error will be "swallowed", which we don't want to do and which wasn't done by bootstrap before. I would still keep the enum to make the behavior more explicit, just removing the option and defaulting to `BehaviorOnFailure::Exit`. What do you think?

File: src/bootstrap/lib.rs
Comment (onur-ozkan) [inline src/bootstrap/lib.rs:1002]: Can this actually return false? I feel like in any failure the program will exit regardless.

Comment (Kobzol) [reply src/bootstrap/lib.rs:1002]: Yes, when a delayed failure command is executd, and `self.fail_fast` is `false`, then this function will return `false`.

**Review by onur-ozkan — APPROVED** [review]

**Comment by Kobzol** [comment]
Hmm, I went over the PR diff and I fail to see anything that could cause CI test failures, as all commands should hopefully be executed same as before. Any ideas?

**Comment by onur-ozkan** [comment]
The CI error log doesn't really help much. What I see is empty stdout and stderr, so I would debug this block:

https://github.com/rust-lang/rust/blob/3bb226fe6c8f42e226db80757c2d59cffcaf50df/src/bootstrap/lib.rs#L1010-L1026

I suspect that the command was probably successful, but we might have made a mistake here.

**Comment by Kobzol** [comment]
I see this error in the CI log:
```
2023-10-12T13:08:18.6284364Z [RUSTC-TIMING] libffi_sys test:false 0.107
2023-10-12T13:08:18.6284660Z [RUSTC-TIMING] libffi test:false 0.729
2023-10-12T13:08:18.6285313Z [0m[1m[38;5;9merror[E0432][0m[0m[1m: unresolved import `std::sync::atomic::AtomicU64`[0m
2023-10-12T13:08:18.6285702Z [0m [0m[0m[1m[38;5;12m--> [0m[0msrc/tools/miri/src/clock.rs:1:25[0m
2023-10-12T13:08:18.6285992Z [0m  [0m[0m[1m[38;5;12m|[0m
2023-10-12T13:08:18.6286407Z [0m[1m[38;5;12m1[0m[0m [0m[0m[1m[38;5;12m|[0m[0m [0m[0muse std::sync::atomic::{AtomicU64, Ordering};[0m
2023-10-12T13:08:18.6287111Z [0m  [0m[0m[1m[38;5;12m| [0m[0m                        [0m[0m[1m[38;5;9m^^^^^^^^^[0m
2023-10-12T13:08:18.6287524Z [0m  [0m[0m[1m[38;5;12m| [0m[0m                        [0m[0m[1m[38;5;9m|[0m
2023-10-12T13:08:18.6287971Z [0m  [0m[0m[1m[38;5;12m| [0m[0m                        [0m[0m[1m[38;5;9mno `AtomicU64` in `sync::atomic`[0m
2023-10-12T13:08:18.6288420Z [0m  [0m[0m[1m[38;5;12m| [0m[0m                        [0m[0m[1m[38;5;9mhelp: a similar name exists in the module: `AtomicU32`[0m
2023-10-12T13:08:18.6288748Z [0m  [0m[0m[1m[38;5;12m|[0m
2023-10-12T13:08:18.6289224Z [0m  [0m[0m[1m[38;5;12m= [0m[0m[1mhelp[0m[0m: consider importing this type alias instead:[0m
2023-10-12T13:08:18.6289637Z [0m          rustc_data_structures::sync::AtomicU64[0m
2023-10-12T13:08:18.6289801Z 
2023-10-12T13:08:18.6290023Z [0m[1mFor more information about this error, try `rustc --explain E0432`.[0m
2023-10-12T13:08:18.6290347Z [RUSTC-TIMING] miri test:false 2.445
2023-10-12T13:08:18.6290720Z [0m[0m[1m[31merror[0m[1m:[0m could not compile `miri` (lib) due to previous error
2023-10-12T13:08:18.6290904Z 
```
This looks weird, like some build option wasn't passed correctly or something, as it cannot find 64-bit atomic types.

**Comment by Kobzol** [comment]
Oh, that's because it *does fail on master*:
https://github.com/rust-lang-ci/rust/actions/runs/6595976804/job/17921554955

So before, the failure of the miri build was ignored, now it isn't. I wonder if the failure is expected or not (?).

**Comment by onur-ozkan** [comment]
> I wonder if the failure is expected or not (?).

I don't think it is?

cc @RalfJung

**Comment by RalfJung** [comment]
Are you sure we were even building Miri on powerpc before this? I certainly wasn't aware that we did, I thought this gets only built on a few select tier 1 targets.

https://github.com/rust-lang/miri/pull/3135 should fix the build failure. I'll get that synced to rustc shortly. But this might happen again in the future since Miri CI doesn't try to do the build on a powerpc host, and I also don't know how we'd even do that.

**Comment by Kobzol** [comment]
Yes, we were building it, because the build also happens on master, and also fails there (it just didn't fail CI before this PR).

The question is whether we shoulf be building it in the first place - perhaps an evem better fix is to disable it completely on powerpc?

**Comment by RalfJung** [comment]
> The question is whether we should be building it in the first place - perhaps an even better fix is to disable it completely on powerpc?

When we say it's a tier 2 target with host tools, which tools does that include? rustc itself, sure, and cargo I assume. But what about clippy and Miri? I don't know. I'm fine with saying that it includes Miri, it just seems hard to test this in Miri CI so we might get a failure when syncing Miri changes to rustc.

Cc @rust-lang/miri

**Comment by onur-ozkan** [comment]
> Sync PR is up: #117044

@rustbot blocked

Let's hold this one until the synchronization PR is merged.

**Comment by saethlin** [comment]
Just speaking as a member of the Miri team, I don't particularly mind running into such issues occasionally. The odds of running into such problems seems very low, and the worst case scenario seems like we just back out a problematic PR from the Miri repo and try to re-land it later.

But at a compiler contributor I'm concerned that this suggests that the target doesn't align with the target tier policy. I am a little surprised I haven't heard of this target blocking a PR before.

**Comment by RalfJung** [comment]
Does our target tier policy require AtomicU64 for host tools?

**Comment by saethlin** [comment]
No, but these are the parts of the policy that stuck out to me when I just read over it:

> the host tools provided for a tier 2 target may include only rustc and cargo, or may include additional tools such as clippy and rustfmt.

No mention of  Miri.

> The host tools must have direct value to people other than the target's maintainers.

The target in question does not even have any target maintainers. Which is typical, but not a significant signal that the target tools are in use.

> The host tools must provide a substantively similar experience as on other targets, subject to reasonable target limitations.

Considering how wildly varying the experience of Miri is among Tier 1 targets (Windows filesystem support), I think it might be fair to say Miri is not one of the host tools covered in the policy.

**Comment by RalfJung** [comment]
> Considering how wildly varying the experience of Miri is among Tier 1 targets (Windows filesystem support), I think it might be fair to say Miri is not one of the host tools covered in the policy.

That's for different targets though, we are largely independent of the host. So the Linux x86 targets should work very well on any host.

> Just speaking as a member of the Miri team, I don't particularly mind running into such issues occasionally. The odds of running into such problems seems very low, and the worst case scenario seems like we just back out a problematic PR from the Miri repo and try to re-land it later. That would be annoying, but probably not take too much of my time or spoons.

Yeah I agree. I don't think I ever had a rustc PR fail due to a tier 2 target host tool build issue, and the fact that this unnecessary use of `AtomicU64` in Miri is the only problem is also a good sign.

**Comment by Kobzol** [comment]
Let's see if the miri update has helped.

@bors try

**Comment by RalfJung** [comment]
From a Miri perspective, we shouldn't ignore build failures. This is probably a leftover from the toolstate days. We shod either build and then stop CI if that fails, or not even attempt to build. So this behavior change seems fine.

But I can understand that you might want to figure out where the behavior change comes from.

**Comment by onur-ozkan** [comment]
Finding the reason of this behaviour and applying the new behaviour in a seperated commit would be great IMO. That way we can revert that specific commit easily instead of reverting the whole implementation if needed.

**Comment by Kobzol** [comment]
Ok I found out the issue. There were three places (finding local git modifications, blessing tests and building a tool) where a failure of the command didn't "crash" bootstrap previously, nor did it delay the failure. This mode wasn't possible in the new command structure, so I have added it, and used it in these three places, to keep the original behavior.

Regarding the tools, most tools built using the `ToolBuild` step are optional, which means that if their build fails, bootstrap will not crash.

File: src/bootstrap/src/utils/exec.rs
Comment (onur-ozkan) [inline src/bootstrap/src/utils/exec.rs:43]:
```suggestion
    pub fn delay_failure(self) -> Self {
        Self { failure_behavior: BehaviorOnFailure::DelayFail, ..self }
    }

    pub fn fail_fast(self) -> Self {
        Self { failure_behavior: BehaviorOnFailure::Exit, ..self }
    }

    pub fn allow_failure(self) -> Self {
        Self { failure_behavior: BehaviorOnFailure::Ignore, ..self }
    }
```

**Review by onur-ozkan — APPROVED** [review]
Thanks! This LGTM.

r=me with or without the suggested minor change

**Pull request merged**
